File: Tyche.Android/appCore/src/main/java/com/felipearpa/appCore/type/BetScore.kt
package com.felipearpa.appCore.type

@JvmInline
value class BetScore(val value: Int) {

    init {
        checkRange()
    }

    private fun checkRange() {
        if (!isValid(value)) {
            throw IllegalArgumentException("Value must be in range [0, 999]")
        }
    }

    override fun toString(): String = value.toString()

    companion object {

        fun isValid(value: Int) = value in 0..999
    }
}

File: Tyche.iOS/Core/Sources/Core/Type/BetScore.swift
private enum BetScoreError: Error {
    case notInRange
}

public struct BetScore {
    public let value: Int
    
    public init?(_ value: Int) {
        if !BetScore.isValid(value: value) {
            return nil
        }
        self.value = value
    }
    
    public static func isValid(value: Int) -> Bool {
        do {
            try checkRange(value: value)
            return true
        } catch {
            return false
        }
    }
    
    private static func checkRange(value: Int) throws {
        if !(0...999 ~= value) {
            throw BetScoreError.notInRange
        }
    }
}
